packages/commonwealth/client/scripts/state.ts

```diff
@@ -234,4 +238,102 @@ const app: IApp = {
   cachedIdentityWidget: null,
 };
 
+// On login: called to initialize the logged-in state, available chains, and other metadata at /api/status
+// On logout: called to reset everything
+export async function initAppState(
+  updateSelectedChain = true,
+  customDomain = null
+): Promise<void> {
+  return new Promise((resolve, reject) => {
+    $.get(`${app.serverUrl()}/status`)
+      .then(async (data) => {
+        app.config.chains.clear();
+        app.config.nodes.clear();
+        app.user.notifications.clear();
+        app.user.notifications.clearSubscriptions();
+
+        data.result.nodes
+          .sort((a, b) => a.id - b.id)
+          .map((node) => {
+            return app.config.nodes.add(NodeInfo.fromJSON(node));
+          });
+
+        data.result.chainsWithSnapshots
+          .filter((chainsWithSnapshots) => chainsWithSnapshots.chain.active)
+          .map((chainsWithSnapshots) => {
+            delete chainsWithSnapshots.chain.ChainNode;
+            return app.config.chains.add(
+              ChainInfo.fromJSON({
+                ChainNode: app.config.nodes.getById(
+                  chainsWithSnapshots.chain.chain_node_id
+                ),
+                snapshot: chainsWithSnapshots.snapshot,
+                ...chainsWithSnapshots.chain,
+              })
+            );
+          });
+
+        app.roles.setRoles(data.result.roles);
+        app.config.notificationCategories =
+          data.result.notificationCategories.map((json) =>
+            NotificationCategory.fromJSON(json)
+          );
+        app.config.invites = data.result.invites;
+        app.config.chainCategories = data.result.chainCategories;
+        app.config.chainCategoryTypes = data.result.chainCategoryTypes;
+
+        // add recentActivity
+        const { recentThreads } = data.result;
+        recentThreads.forEach(({ chain, count }) => {
+          app.recentActivity.setCommunityThreadCounts(chain, count);
+        });
+
+        // update the login status
+        updateActiveUser(data.result.user);
+        app.loginState = data.result.user
+          ? LoginState.LoggedIn
+          : LoginState.LoggedOut;
+
+        if (app.loginState === LoginState.LoggedIn) {
+          console.log('Initializing socket connection with JTW:', app.user.jwt);
+          // init the websocket connection and the chain-events namespace
+          app.socket.init(app.user.jwt);
+          app.user.notifications.refresh().then(() => m.redraw());
+        } else if (
+          app.loginState === LoginState.LoggedOut &&
+          app.socket.isConnected
+        ) {
+          // TODO: create global deinit function
+          app.socket.disconnect();
+        }
```

Comment:
I think we should make this into a helper and reduce the size of the initAppState function